### PR TITLE
Enable arbitrary euler angle for Mag rotation

### PR DESCRIPTION
**Pull request by junwoo091400**
## Describe problem solved by this pull request
Previously, the rotations for the magnetometer was confined to MAVLink defined [MAV_SENSOR_ORIENTATION](https://mavlink.io/en/messages/common.html#MAV_SENSOR_ORIENTATION), which doesn't allow setting non 45-degree multiple sensor orientation.

This was a limiting factor for drones with magnetometer that has Magnetometer installed in Euler angles of something like 37 degrees, 58 degrees, etc.

## Describe your solution
Enables setting arbitrary euler angle for magnetometer orientation, which would allow users to set angles like 37 degrees in yaw, which is useful for Alta X for example

### Added parameters
Parameters that defines the Euler angle for the orientation the Magnetometer is at were added:

- CAL_MAG${i}_ROLL
- CAL_MAG${i}_PITCH
- CAL_MAG${i}_YAW

## Describe possible alternatives
Quaternion?

## Test data / coverage
TODO

## TODO
- [ ] Add QGC UI to dynamically visualize this Euler angle setting for custom rotation

## Discussion
Currently, due to bringing in the MAV_SENSOR_ROTATION_CUSTOM enum, which breakes the sequence of numbers, and bumps up the value to 100, the `ROTATION_MAX` can no longer be used to calculate the number of rotations we support in PX4.

You can read more about enum members number count issue [here](https://stackoverflow.com/questions/14989274/is-it-possible-to-determine-the-number-of-elements-of-a-c-enum-class).

Would there be a good solution for this?

### CODE REVIEW

File: src/lib/conversion/rotation.h
Comment (dagar) [inline src/lib/conversion/rotation.h:96]: This gets a bit weird with things like mag cal auto rotation iteration through enum Rotation. What about just keeping `ROTATION_MAX = 41` right after that last valid rotation enum that isn't custom?

Comment (junwoo091400) [reply src/lib/conversion/rotation.h:96]: Ahhh that is a nice way to solve this problem! Thank you!

**Comment by junwoo091400** [comment]
Review addressed :wink:

**Comment by junwoo091400** [comment]
Any extra comments about this?

**Comment by junwoo091400** [comment]
FYI, betaflight has it as well
https://github.com/betaflight/betaflight/blob/5e7bca15768056a818ea867a8bd08e825666fde7/src/main/sensors/compass.c#L363

**Comment by junwoo091400** [comment]
Related: https://github.com/betaflight/betaflight/commit/980df1536fb56c6cab369b360aa4580cb9b102d3

File: src/lib/conversion/rotation.h
Comment (bresch) [inline src/lib/conversion/rotation.h:94]:
```suggestion
	ROTATION_MAX,
```
This is a good example to show that comments rot quickly; even in the same PR they were created.

**Review by bresch — CHANGES_REQUESTED** [review]

File: src/lib/conversion/rotation.h
Comment (junwoo091400) [reply src/lib/conversion/rotation.h:94]: Oh yes, good to remove the comment, but I meant that it should be placed at the end of the 'non-custom' rotation (in case someone would place it after the ROTATION_CUSTOM :exploding_head:)

**Comment by junwoo091400** [comment]
Maybe having a QGC UI would unblock this from being merged? Something like this: https://www.youtube.com/watch?v=mENP56CmeVw. But it requires OpenGL & would be quite a bit of an effort!

Or, alternatively, we can also just update the doc & with mention a pre-existing online tool: https://danceswithcode.net/engineeringnotes/rotations_in_3d/demo3D/rotations_in_3d_tool.html

**Comment by bresch** [comment]
What's blocking this PR? The users probably don't need a tool to represent the orientations if they're already supposed to be able to select the orientation from a list.

**Comment by junwoo091400** [comment]
> What's blocking this PR? The users probably don't need a tool to represent the orientations if they're already supposed to be able to select the orientation from a list.

Nothing is really blocking it. Got it, then I think we can get it in :thinking:, with doc changes.

I could maybe add a unit test for this?

File: src/lib/conversion/rotation.h
Comment (junwoo091400) [reply src/lib/conversion/rotation.h:94]: I wanted to emphasize that this shouldn't be placed after the ROTATION_CUSTOM. Would this comment not be necessary?

**Comment by junwoo091400** [comment]
Rebased on `main` and added comment / logic about the sensor level adjustment

File: src/lib/sensor_calibration/Magnetometer.cpp
Comment (junwoo091400) [inline src/lib/sensor_calibration/Magnetometer.cpp:142]: @bresch Is this an accurate description?

File: src/lib/sensor_calibration/Magnetometer.hpp
Comment (junwoo091400) [inline src/lib/sensor_calibration/Magnetometer.hpp:122]: @dagar I in general notice that the frame rotation matrices / biases don't have a clear description, so it's unclear when & where the bias should be applied (before or after the rotation?).

As you mentioned in [this comment](https://github.com/PX4/PX4-Autopilot/pull/20543#issue-1434820706), I think partial solution would be to be more verbose in naming / commenting on the bias / rotation matrices on exactly what they are supposed to do.

Thoughts?

**Comment by hamishwillee** [comment]
FMI is this a real problem? I mean how often does someone actually feel limited by the inability to set a 37 degree angle on the magnetometer?

W.r.t. MAVLink rather than an 8 bit number to send the current info you're talking 3 floats - that's 12 times more information - with all the corresponding changes to the messages etc. Is it worth it/has anyone thought this through?

I don't "care" per se, just if you put a PR in to MAVLink to support this it needs justification.

**Comment by junwoo091400** [comment]
> FMI is this a real problem? I mean how often does someone actually feel limited by the inability to set a 37 degree angle on the magnetometer?

Here I showcase one example, my X500 V2. This model is *supposed to be built by mounting the GPS on it's dedicated mount on the panel (flat horizontal plate sticking out), but due to the concern of having too much vibration induced on that panel, we mounted the GPS on the rigid frame.

And since I already had taped the GPS on the mount before attaching it to the frame, and the yaw 23 degrees or so (that's all the adjustment we needed) wasn't supported in mag rotation, I had to un-tape the whole GPS and tape it again, which was quite ironic since taping it back on isn't even more accurate (It will never be really ROTATION_NONE).

This wasn't however the only example we had, but other products even needed a hardware mount to orient the mag in a *supported rotation enum, which isn't scalable for a mass produced vehicles, so that's why this project was initiated.

Hope this answers your question!

**Image1: How the GPS was re-taped to be oriented forward (ROTATION_NONE)**


**Image2: How the GPS was mounted (it's a non-standard mounting method, hence it resulted in yaw not being 0 degrees)**

**Comment by junwoo091400** [comment]
> I'm not opposed - but if you put a PR in to MAVLink (and hence to QGC) to support this more deeply in the UI it needs more justification than the statement that it's a problem.

I don't think this needs to be related to MAVLink, this is just about supporting the custom rotation in PX4 :thinking:. Or am I missing something?

**Comment by junwoo091400** [comment]
I don't think anything is preventing this from being merged. And we discussed in the dev call as well, where general consensus that this is a good addition to solve a pain of having to mount everything as specified via enum (or implementing custom mavlink enum for custom rotation): https://discuss.px4.io/t/px4-dev-call-november-09-2022/29606#p3-arbitrary-mag-rotation-junwoo-12

@hamishwillee I hope this is clear for you!

File: src/lib/sensor_calibration/Magnetometer.hpp
Comment (tstastny) [reply src/lib/sensor_calibration/Magnetometer.hpp:122]: > be more verbose in naming / commenting on the bias / rotation matrices

both. please.

@junwoo091400  since you specified this rotation at least in a comment, i don't want to drag out this PR any longer. perhaps you could make an issue though, and we can try to get a discussion started on defining a convention? i.e. what specific naming for rotation matrices and quaternions should we try to be consistent with throughout the code base?

**Review by tstastny — APPROVED** [review]

**Comment by tstastny** [comment]
> Additional discussion from dev call:
> - Wipe out the CAL_* parameters? (Since it's not vehicle specific)
> - And have SENS_* stuff should be preserved (since it's board-specific, and shouldn't be changed)

@dagar @bresch - is this a blocker for the pr? if cal and sens are used as airframe vs board, probably @junwoo091400 it would make sense to change, but i defer to @bresch and @dagar

**Comment by junwoo091400** [comment]
> > Additional discussion from dev call:
> > 
> > * Wipe out the CAL_* parameters? (Since it's not vehicle specific)
> > * And have SENS_* stuff should be preserved (since it's board-specific, and shouldn't be changed)
> 
> @dagar @bresch - is this a blocker for the pr? if cal and sens are used as airframe vs board, probably @junwoo091400 it would make sense to change, but i defer to @bresch and @dagar

This is external compass, so it's airframe specifc and makes sense to have the `CAL_` prefix (as in this PR). Also, the discussion was something that me and daniel were lightly talking about, so I may have taken the note down too formally :laughing

Maybe @bresch you could confirm whether `CAL_*` prefix makes sense? I def think it does :thinking:

File: src/lib/sensor_calibration/Magnetometer.hpp
Comment (junwoo091400) [reply src/lib/sensor_calibration/Magnetometer.hpp:122]: Issue created: https://github.com/PX4/PX4-Autopilot/issues/20639

**Comment by junwoo091400** [comment]
Shall we get this in? @dagar @bresch

**Comment by hamishwillee** [comment]
Any ETA on this?

**Review by mcsauder — APPROVED** [review]
I like your const correctness and I'd like to see this added to imu accels/gyros as well!  Nice work!

**Comment by junwoo091400** [comment]
Note: I need to rebase to take into consideration of https://github.com/PX4/PX4-Autopilot/pull/20723

**Comment by tstastny** [comment]
@junwoo091400 will you have time to wrap this up this week? if not - we may be able to put someone else on it.

**Comment by junwoo091400** [comment]
> what else is remaining beside rebase?

Just the renaming of the parameters prefix, I can get it settled certainly by 22nd of June (end of Exams), would that be too late?

File: src/modules/sensors/module.yaml
Comment (dagar) [inline src/modules/sensors/module.yaml:398]: This is probably enough spacing, but just to be safe maybe we should set this to a much larger magic number.

How about either UINT8_MAX (255) or even INT32_MAX? Currently the underlying parameter is actually an int32_t, but could be optimized in the future.

Comment (bresch) [reply src/modules/sensors/module.yaml:398]: I think we should eventually only use the "custom rotation", so this parameter will be completely removed

**Comment by bresch** [comment]
replaced by https://github.com/PX4/PX4-Autopilot/pull/22000

**Pull request closed without merging**
